lib/internal/iterator.js

```diff
@@ -0,0 +1,40 @@
+'use strict';
+
+import isArrayLike from 'lodash/isArrayLike';
+import keys from 'lodash/keys';
+import isMap from 'lodash/isMap';
+
+var iteratorSymbol = typeof Symbol === 'function' && Symbol.iterator;
+
+export default function iterator(coll) {
+    var i = -1;
+    var len;
+    if (isArrayLike(coll)) {
+        len = coll.length;
+        return function next() {
+            i++;
+            return i < len ? {value: coll[i], key: i} : null;
+        };
+    }
+
+    if (iteratorSymbol && coll[iteratorSymbol]) {
+        var iterate = coll[iteratorSymbol]();
+        var map = isMap(coll);
+
+        return function next() {
+            var item = iterate.next();
+            if (item.done)
+                return null;
+            i++;
+            return map ? {value: item.value[1], key: item.value[0]} : {value: item.value, key: i};
```

Comment:
I think this is a valid thing to do.  I think it's odd to iterate over a map and get `[key, val]` as the value with an arbitrary numeric index.